packages/autocomplete-plugin-tags/src/createTagsPlugin.tsx

```diff
@@ -0,0 +1,160 @@
+/** @jsx createElement */
+import { BaseItem, PluginSubscribeParams } from '@algolia/autocomplete-core';
+import {
+  AutocompletePlugin,
+  AutocompleteSource,
+  AutocompleteState,
+} from '@algolia/autocomplete-js';
+import { noop } from '@algolia/autocomplete-shared';
+
+import { createTags, OnTagsChangeParams } from './createTags';
+import type { BaseTag, Tag } from './types';
+
+type OnChangeParams<TTag> = PluginSubscribeParams<any> &
+  OnTagsChangeParams<TTag>;
+
+type TagsPluginData<TTag> = {
+  /**
+   * Returns the current list of tags.
+   *
+   * @link https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-plugin-tags/createtagsplugin/#param-tags
+   */
+  tags: Array<Tag<TTag>>;
+  /**
+   * Adds tags to the list.
+   *
+   * @link https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-plugin-tags/createtagsplugin/#param-addtags
+   */
+  addTags: (tags: Array<BaseTag<TTag>>) => void;
+  /**
+   * Sets the list of tags.
+   *
+   * @link https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-plugin-tags/createtagsplugin/#param-settags
+   */
+  setTags: (tags: Array<BaseTag<TTag>>) => void;
+};
+
+export type CreateTagsPluginParams<
+  TTag extends Record<string, any>,
+  TItem extends BaseItem
+> = {
+  /**
+   * A set of initial tags to pass to the plugin.
+   *
+   * @link https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-plugin-tags/createtagsplugin/#param-initialtags
+   */
+  initialTags?: Array<BaseTag<TTag>>;
+  /**
+   * A function to specify what sources the plugin should subscribe to. The plugin adds a tag when selecting an item from these sources.
+   *
+   * @link https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-plugin-tags/createtagsplugin/#param-gettagssubscribers
+   */
+  getTagsSubscribers?(): Array<{
+    sourceId: string;
+    getTag(params: { item: TItem }): BaseTag<TTag>;
+  }>;
+  /**
+   * A function to transform the returned tags source.
+   *
+   * @link https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-plugin-tags/createtagsplugin/#param-transformsource
+   */
+  transformSource?(params: {
+    source: AutocompleteSource<Tag<TTag>>;
+    state: AutocompleteState<Tag<TTag>>;
+  }): AutocompleteSource<Tag<TTag>> | undefined;
```

Comment:
Our other plugins don't accept `undefined` as the return of `transformSource`, why do we need it here?

Reply:
That's because in our other plugins we always display a source. In the Tags plugin we display a source bu default bit if the user wants to render it manually (e.g. in the searchbox) they can return undefined here to discard the default source.

Reply:
IIRC you can return `null`, which we accept as a type.

Reply:
[No we don't](https://github.com/algolia/autocomplete/blob/next/packages/autocomplete-js/src/types/AutocompleteOptions.ts#L31-L33), which is why [I changed it](https://github.com/algolia/autocomplete/commit/a1e0fc20ad73ef9a4be22bb6e695f362d54d10af). `null` semantically makes a lot more sense but I figured this was a design choice to expect `undefined`.

I'm happy to go back to `null`, but we either need to add some logic to return `undefined`, or accept `null` as a valid `GetSources` return type.